### PR TITLE
Port QueryEngine

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This ports the query engine changes for indexed query execution from Android.

### CODE REVIEW

File: packages/firestore/src/core/query.ts [hunk 1]
```diff
@@ -155,7 +155,7 @@ export function asCollectionQueryAtPath(
  * Returns true if this query does not specify any query constraints that
  * could remove results.
  */
-export function matchesAllDocuments(query: Query): boolean {
+export function queryMatchesAllDocuments(query: Query): boolean {
```
Comment (schmidt-sebastian) [inline packages/firestore/src/core/query.ts:158]: Changed to match naming convection: Classs+MethodName.

File: packages/firestore/test/unit/local/query_engine.test.ts [hunk 2]
```diff
@@ -307,8 +387,9 @@ describe('QueryEngine', () => {
 
     // Add a query mapping for a document that matches, but that sorts below
     // another document due to a pending write.
-    await addDocument(PENDING_MATCHING_DOC_A);
-    await persistQueryMapping(PENDING_MATCHING_DOC_A.key);
+    await addDocument(MATCHING_DOC_A);
+    await addMutation(patchMutation('coll/a', { order: 1 }));
```
Comment (schmidt-sebastian) [inline packages/firestore/test/unit/local/query_engine.test.ts:391]: These changes are to support IndexedDb persistence. We cannot store a document with mutations in indexeddb so we instead have to store a document and its mutation.

File: packages/firestore/src/local/indexeddb_index_manager.ts [hunk 3]
```diff
@@ -972,6 +972,16 @@ export class IndexedDbIndexManager implements IndexManager {
     }
     return ranges;
   }
+
+  getMinOffset(
+    transaction: PersistenceTransaction,
+    target: Target
+  ): PersistencePromise<IndexOffset> {
+    // TODO(orqueries): Get the minimum offset for all subqueries
+    return this.getFieldIndex(transaction, target).next(index =>
```
Comment (wu-hui) [inline packages/firestore/src/local/indexeddb_index_manager.ts:981]: So this is required by or query, and left to implement when we port or queries? Just to clarify.

File: packages/firestore/src/local/query_engine.ts [hunk 4]
```diff
@@ -74,15 +110,122 @@ export class QueryEngine {
...
+    query: Query
+  ): PersistencePromise<DocumentMap | null> {
+    if (!INDEXING_ENABLED) {
+      return PersistencePromise.resolve<DocumentMap | null>(null);
+    }
+
+    if (queryMatchesAllDocuments(query)) {
+      // Don't use indexes for queries that can be executed by scanning the
+      // collection.
+      return PersistencePromise.resolve<DocumentMap | null>(null);
```
Comment (wu-hui) [inline packages/firestore/src/local/query_engine.ts:146]: // No need to use indexes if the query requires full collection scan anyways.

File: packages/firestore/src/local/query_engine.ts [hunk 5]
```diff
@@ -74,15 +110,122 @@ export class QueryEngine {
...
+                    if (
+                      (hasLimitToFirst(query) || hasLimitToLast(query)) &&
+                      this.needsRefill(
+                        query.limitType,
+                        previousResults,
+                        sortedKeys,
+                        offset.readTime
+                      )
+                    ) {
+                      return PersistencePromise.resolve<DocumentMap | null>(
```
Comment (wu-hui) [inline packages/firestore/src/local/query_engine.ts:198]: Can we add a comment to explain why we are returning null here?

File: packages/firestore/test/unit/local/query_engine.test.ts [hunk 6]
```diff
@@ -85,11 +106,47 @@ class TestLocalDocumentsView extends LocalDocumentsView {
   }
 }
 
-describe('QueryEngine', () => {
+describe('MemoryQueryManager', async () => {
```
Comment (wu-hui) [inline packages/firestore/test/unit/local/query_engine.test.ts:109]: QueryEngine ?

Same below.

File: packages/firestore/test/unit/local/query_engine.test.ts [hunk 7]
```diff
@@ -403,7 +485,85 @@ describe('QueryEngine', () => {
...
+      3,
+      LimitType.First
+    );
+    const results = await expectOptimizedCollectionQuery(() =>
+      runQuery(q, SnapshotVersion.min())
+    );
+
+    verifyResult(results, [doc2]);
+  });
+}
```
Comment (wu-hui) [inline packages/firestore/test/unit/local/query_engine.test.ts:566]: We should also add a test for refilling, unless there is already one. It's late for me so I did not bother to check.

File: packages/firestore/src/local/query_engine.ts (hunk 5 above)
Comment (wu-hui) [reply packages/firestore/src/local/query_engine.ts]: Thanks!

**Review by wu-hui — APPROVED** [review]

**Pull request merged**